mrmustard/physics/representations.py

```diff
@@ -0,0 +1,187 @@
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from mrmustard import math
+from mrmustard.physics import bargmann
+from mrmustard.physics.ansatze import Ansatz, PolyExpAnsatz
+from mrmustard.utils.typing import Batch, ComplexMatrix, ComplexTensor, ComplexVector
+
+
+class Representation:
+    _contract_idxs: tuple[int, ...] = ()
+
+    def from_ansatz(self, ansatz: Ansatz) -> Ansatz:
+        raise NotImplementedError
+
+    def __eq__(self, other):
+        return self.ansatz == other.ansatz
+
+    def __add__(self, other):
+        return self.from_ansatz(self.ansatz + other.ansatz)
+
+    def __sub__(self, other):
+        return self.from_ansatz(self.ansatz - other.ansatz)
+
+    def __mul__(self, other):
+        try:
+            return self.from_ansatz(self.ansatz * other.ansatz)
+        except AttributeError:
+            return self.from_ansatz(self.ansatz * other)
+
+    def __rmul__(self, other):
+        return self.__mul__(other)
+
+    def __truediv__(self, other):
+        try:
+            return self.from_ansatz(self.ansatz / other.ansatz)
+        except AttributeError:
+            return self.from_ansatz(self.ansatz / other)
+
+    def __rtruediv__(self, other):
+        return self.from_ansatz(other / self.ansatz)
+
+    def __and__(self, other):
+        return self.from_ansatz(self.ansatz & other.ansatz)
+
+
+class Bargmann(Representation):
+    r"""Fock-Bargmann representation of a broad class of quantum states,
+    transformations, measurements, channels, etc.
+    The ansatz available in this representation is a linear combination of
+    exponentials of bilinear forms with a polynomial part:
+    .. math::
+        F(z) = sum_i poly_i(z) exp(z^T A_i z / 2 + z^T b_i)
+
+    This function allows for vector space operations on Bargmann objects including linear combinations,
+    outer product, and inner product. The inner product is defined as the contraction of two
+    Bargmann objects across marked indices. This can also be used to contract existing indices
+    in one Bargmann object, e.g. to implement the partial trace.
+
+    Args:
+        A (Batch[ComplexMatrix]): batch of quadratic coefficient A_i
+        b (Batch[ComplexVector]): batch of linear coefficients b_i
+        c (Batch[ComplexTensor]): batch of arrays c_i (default: [1.0])
+    """
+
+    def __init__(
+        self,
+        A: Batch[ComplexMatrix],
+        b: Batch[ComplexVector],
+        c: Batch[ComplexTensor] = [1.0],
+    ):
+        self.ansatz = PolyExpAnsatz(A, b, c)
+        self._name = "Bargmann"
```

Comment:
Do we need the name? We can just use isinstance to know if this representation is bargmann